File: stack/155_min_stack.py
"""
https://leetcode.com/problems/min-stack/description/

two stacks

else:

if element < min
push 2*element - min
min = element

pop popped_element < min
new_min 2*min - popped_element
return min

explanation on ipad

"""


class MinStack:

    def __init__(self):
        self.stack = []
        self.min = float('-inf')

    def push(self, val: int) -> None:
        if len(self.stack) == 0:
            self.stack.append(val)
            self.min = val
        else:
            if val > self.min:
                self.stack.append(val)
            else:
                push_val = 2 * val - self.min
                self.min = val
                self.stack.append(push_val)

    def pop(self) -> None:
        element = self.stack.pop()
        if element < self.min:
            a = self.min
            self.min = 2 * self.min - element
            return a

        return element

    def top(self) -> int:
        element = self.stack[-1]
        if element < self.min:
            return self.min

        return element

    def getMin(self) -> int:
        return self.min

# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

File: stack/155_min_stack.go
package main

import "math"

type MinStack struct {
	stack   []int
	minElem int
}

func Constructor() MinStack {
	minStack := MinStack{
		stack:   make([]int, 0),
		minElem: math.MaxInt,
	}

	return minStack
}

func (this *MinStack) Push(val int) {
	// zero case should be handled separately
	if len(this.stack) == 0 {
		this.stack = append(this.stack, val)
		this.minElem = val
	} else {
		if val > this.minElem {
			this.stack = append(this.stack, val)
		} else {
			this.stack = append(this.stack, 2*val-this.minElem)
			this.minElem = val
		}
	}
}

func (this *MinStack) Pop() {
	top_elem := this.stack[len(this.stack)-1]
	if top_elem < this.minElem {
		this.minElem = 2*this.minElem - top_elem
	}

	this.stack = this.stack[:len(this.stack)-1]

}

func (this *MinStack) Top() int {
	top_elem := this.stack[len(this.stack)-1]
	elem := top_elem
	if top_elem < this.minElem {
		elem = this.minElem
	}

	return elem
}

func (this *MinStack) GetMin() int {
	return this.minElem
}

/**
 * Your MinStack object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Push(val);
 * obj.Pop();
 * param_3 := obj.Top();
 * param_4 := obj.GetMin();
 */
